ext/dom/xpath_callbacks.c

```diff
@@ -0,0 +1,504 @@
+		}
+		efree(params);
+	}
+}
+
+static zend_result php_dom_xpath_callback_dispatch(php_dom_xpath_callbacks *xpath_callbacks, php_dom_xpath_callback_ns *ns, xmlXPathParserContextPtr ctxt, zval *params, uint32_t param_count, const char *function_name, size_t function_name_length)
+{
+    zval callback_retval;
+
+	if (UNEXPECTED(ns == NULL)) {
+		zend_throw_error(NULL, "No callbacks were registered");
+		return FAILURE;
+	}
+
+	if (ns->mode == PHP_DOM_REG_FUNC_MODE_ALL) {
+		zend_fcall_info fci;
+		fci.size = sizeof(fci);
+		fci.object = NULL;
+		fci.retval = &callback_retval;
+		fci.param_count = param_count;
+		fci.params = params;
+		fci.named_params = NULL;
+		ZVAL_STRINGL(&fci.function_name, function_name, function_name_length);
+
+		zend_result result = zend_call_function(&fci, NULL);
+		zend_string_release_ex(Z_STR(fci.function_name), false);
+		if (UNEXPECTED(result == FAILURE)) {
+			return FAILURE;
+		}
+	} else {
+		ZEND_ASSERT(ns->mode == PHP_DOM_REG_FUNC_MODE_SET);
+
+		zval *fcc_wrapper = zend_hash_str_find(&ns->functions, function_name, function_name_length);
+		if (fcc_wrapper) {
+			zend_call_known_fcc(Z_PTR_P(fcc_wrapper), &callback_retval, param_count, params, NULL);
+		} else {
+			zend_throw_error(NULL, "No callback handler \"%s\" registered", function_name);
+			return FAILURE;
+		}
+	}
+
+	if (Z_TYPE(callback_retval) != IS_UNDEF) {
+		if (Z_TYPE(callback_retval) == IS_OBJECT && instanceof_function(Z_OBJCE(callback_retval), dom_node_class_entry)) {
+			xmlNode *nodep;
+			dom_object *obj;
+			if (xpath_callbacks->node_list == NULL) {
+				xpath_callbacks->node_list = zend_new_array(0);
+			}
+			Z_ADDREF_P(&callback_retval);
+			zend_hash_next_index_insert_new(xpath_callbacks->node_list, &callback_retval);
+			obj = Z_DOMOBJ_P(&callback_retval);
+			nodep = dom_object_get_node(obj);
+			valuePush(ctxt, xmlXPathNewNodeSet(nodep));
+		} else if (Z_TYPE(callback_retval) == IS_FALSE || Z_TYPE(callback_retval) == IS_TRUE) {
+			valuePush(ctxt, xmlXPathNewBoolean(Z_TYPE(callback_retval) == IS_TRUE));
+		} else if (Z_TYPE(callback_retval) == IS_OBJECT) {
+			zend_type_error("Only objects that are instances of DOMNode can be converted to an XPath expression");
+			zval_ptr_dtor(&callback_retval);
+			return FAILURE;
+		} else {
+			zend_string *str = zval_get_string(&callback_retval);
+			valuePush(ctxt, xmlXPathNewString((xmlChar *) ZSTR_VAL(str)));
+			zend_string_release_ex(str, 0);
+		}
+		zval_ptr_dtor(&callback_retval);
+	}
+
+	return SUCCESS;
+}
+
+PHP_DOM_EXPORT zend_result php_dom_xpath_callbacks_call_php_ns(php_dom_xpath_callbacks *xpath_callbacks, xmlXPathParserContextPtr ctxt, int num_args, php_dom_xpath_nodeset_evaluation_mode evaluation_mode, dom_object *intern, php_dom_xpath_callbacks_proxy_factory proxy_factory)
+{
+	zend_result result = FAILURE;
+
+	if (UNEXPECTED(num_args == 0)) {
+		zend_throw_error(NULL, "Function name must be passed as the first argument");
+		goto cleanup_no_obj;
+	}
```

Comment:
Is this an API usage failure? As this seems like it should only ever be possible to hit this case if C code does something wrong? Which maybe means this should be an assertion? Or am I missing something?

Reply:
Yes this is possible to hit by using a wrong XSL template that does: `php:function()` in its xpath expression instead of `php:function('function_name')`.